Add localisation for overlay headers

kj415j45 opened the pull request:
Depends on:
- [x] #13911
- [x] ppy/osu-web#7937
- [x] osu-resources strings update and bump

kj415j45 committed:
Add localisation for WikiHeader

Co-authored-by: huoyaoyuan <huoyaoyuan@hotmail.com>

kj415j45 committed:
Add localisation for NewsHeader

Co-authored-by: huoyaoyuan <huoyaoyuan@hotmail.com>

bdach commented:
Please don't open incomplete PRs. If you wish to reserve a task to work on, open an issue for it instead and note that you will be working on it.

bdach commented on osu.Game/Overlays/News/NewsHeader.cs:
```diff
@@ -4,20 +4,24 @@
     public class NewsHeader : BreadcrumbControlOverlayHeader
     {
-        private const string front_page_string = "frontpage";
+        public static LocalisableString FrontPageString => NewsStrings.FrontPageString;
+        public static LocalisableString HeaderTitle => NewsStrings.HeaderTitle;
+        public static LocalisableString HeaderDescription => NewsStrings.HeaderDescription;
```
Should be `public static readonly`.

kj415j45 replied:
`readonly` is invalid for this.
this code format is the same as other localisation PR.

bdach replied:
Fair, misread.

bdach commented:
There should be no `osu-resources` changes required, as I understand the strings being translated are already translated on `osu-web`. Those are [already available for consumption](https://github.com/ppy/osu-resources/tree/master/osu.Game.Resources/Localisation/Web).

kj415j45 commented:
> There should be no `osu-resources` changes required (except for maybe `NewsStrings.HeaderDescription`), as I understand the strings being translated are already translated on `osu-web`. Those are [already available for consumption](https://github.com/ppy/osu-resources/tree/master/osu.Game.Resources/Localisation/Web?rgh-link-date=2021-07-17T12%3A22%3A17Z).

Will check if there're available strings in web later

peppy commented:
As already said, please remove the remaining tasks from this PR and do them separately.

kj415j45 committed:
Optimize code style

Co-authored-by: frenzibyte <frenzibyte@gmail.com>
Co-authored-by: bdach <dach.bartlomiej@gmail.com>

frenzibyte commented on osu.Game/Overlays/Wiki/WikiHeader.cs:
```diff
@@ -5,14 +5,18 @@
     {
-        private const string index_page_string = "index";
         private const string index_path = "Main_Page";
+        public static LocalisableString IndexPageString => WikiStrings.IndexPageString;
+        public static LocalisableString HeaderTitle => WikiStrings.HeaderTitle;
+        public static LocalisableString HeaderDescription => WikiStrings.HeaderDescription;
```
I don't really see why this is declared twice, when you could just prepend `WikiStrings` in every usage, making it read even better. (same applies to `NewsHeader`)

kj415j45 replied:
Full PR Reference:

[1] [LocalisableString usage](https://github.com/peppy/osu/blob/master/osu.Game/Overlays/ChatOverlay.cs#L38-L39)
```csharp
        public LocalisableString Title => ChatStrings.HeaderTitle;
        public LocalisableString Description => ChatStrings.HeaderDescription;
```

[2] [Code format of `*Strings.cs`](https://github.com/ppy/osu/pull/12495#issue-619104156)
```csharp
/// <summary>
/// "Fallback / default text"
/// </summary>
public static LocalisableString LookupKey => new TranslatableString(getKey("lookup_key"), "Fallback / default text");
```

p.s. will remove `static` from `*Header.cs` to match [1]

bdach replied:
The difference there is that `ChatOverlay` was *forced* to redeclare those members, because of implementing `INamedOverlayComponent`.

That said this PR has several places where the same string is referred to multiple times, so I'm not overly fussed about this myself.

kj415j45 replied:
I see.

kj415j45 committed:
Add localisation for ChangelogHeader

kj415j45 committed:
Add localisation for RankingOverlay

kj415j45 committed:
Move strings together

kj415j45 committed:
Add localisation for RankingOverlayHeader

kj415j45 committed:
Add localisation for BeatmapListingHeader

kj415j45 committed:
Add localisation for DashboardOverlayHeader

frenzibyte commented on osu.Game/Overlays/News/NewsHeader.cs:
```diff
@@ -4,20 +4,22 @@
     {
-        private const string front_page_string = "frontpage";
-
+        public LocalisableString FrontPageString => osu.Game.Resources.Localisation.Web.NewsStrings.IndexTitleInfo;
```
should be static, also should have a newline between a property and a field.
```suggestion
        public static LocalisableString FrontPageString => osu.Game.Resources.Localisation.Web.NewsStrings.IndexTitleInfo;

```

frenzibyte commented on osu.Game/Overlays/Wiki/WikiHeader.cs:
```diff
@@ -5,14 +5,16 @@
     {
-        private const string index_page_string = "index";
         private const string index_path = "Main_Page";
+        public LocalisableString IndexPageString => LayoutStrings.HeaderHelpIndex;
```
static + newline between field and property.
```suggestion

        public static LocalisableString IndexPageString => LayoutStrings.HeaderHelpIndex;
```

frenzibyte reviewed (changes requested):
If you're planning for full overlay header localisation, `BeatmapSetOverlay`, `ChatOverlay`, and `UserProfileOverlay` should be localised as well.

While I was checking with osu!web, the strings used for the overlay titles are all stored on `PageTitleStrings`, but some strings used here don't look from there:
https://github.com/ppy/osu/blob/f80b18377775e3bf52b4370d0bd7459dc383112c/osu.Game/Overlays/Changelog/ChangelogHeader.cs#L120

https://github.com/ppy/osu/blob/f80b18377775e3bf52b4370d0bd7459dc383112c/osu.Game/Overlays/Rankings/RankingsOverlayHeader.cs#L36

These use the strings from the web's navigation menu items, displaying as the overlay's title on the header.

I presume that this is done to look theoretically correct on the toolbar when hovering over the overlay buttons (as a somewhat analogue to web's navigation menu), but that's weird for overlay header titles, and has already caused an issue which has been dealt with weirdly:

https://github.com/ppy/osu/blob/1b4bff0d9fa52dec5b18f33158b4201bcb1ed2e2/osu.Game/Overlays/Wiki/WikiHeader.cs#L78-L79

(now the overlay title uses web's navigation menu item text, and the overlay description uses the page title which is supposed to be for the overlay title..??)

I think what should happen here is adding a separate `ShortTitle` property on `INamedOverlayComponent`s for the toolbar to use instead, and let `WikiOverlay` set that to `LayoutStrings.MenuHelpGetWiki`. 

Then let all overlays use the correct `PageTitleStrings` for `Title`.

kj415j45 commented on osu.Game/Localisation/HeaderDescriptionStrings.cs:
```diff
@@ -0,0 +1,39 @@
+
+namespace osu.Game.Localisation
+{
+    public static class HeaderDescriptionStrings
```
```csharp
        /// <inheritdoc cref="PageTitleStrings.MainBeatmapsetsControllerIndex"/>
        public static LocalisableString BeatmapListingTitle => PageTitleStrings.MainBeatmapsetsControllerIndex;
```
Confirmed that this doesn't work on current version of osu-localisation-analyser. The reason is missing `getKey()` method fot it to identify the key properly.

frenzibyte replied:
It's supposed to just be ignored by the generator, not sure what you mean?

kj415j45 committed:
Apply suggestion from code review

Co-authored-by: frenzibyte <frenzibyte@gmail.com>

kj415j45 committed:
symbol renaming

Co-authored-by: frenzibyte <frenzibyte@gmail.com>

kj415j45 commented:
Have to admit that I do mis-match string of `Changelog`'s title.

---

But the `Ranking(s)`' title is another tiny problem:
`LayoutStrings.MenuRankingsDefault` match current in-game string `rankings`. however, `PageTitleStrings.MainRankingControllerDefault` which is suppose to be used is `ranking`, missing an `s`.

frenzibyte commented:
Not sure why they're different, but `PageTitleStrings.MainRankingControllerDefault` will be used for the rankings overlay header title, and `PageTitleStrings.MainRankingControllerDefault` will be used for the toolbar name, after I get to adding `ShortTitle` to `INamedOverlayComponent` (or a better name because it doesn't make sense now).

bdach commented on osu.Game/Overlays/Rankings/RankingsOverlayHeader.cs:
```diff
@@ -29,18 +33,44 @@ private class RankingsTitle : OverlayTitle
         }
     }
 
+    [LocalisableEnum(typeof(RankingsScopeEnumLocalisationMapper))]
     public enum RankingsScope
     {
         Performance,
         Spotlights,
         Score,
         Country
     }
+
+    public class RankingsScopeEnumLocalisationMapper : EnumLocalisationMapper<RankingsScope>
+    {
+        public override LocalisableString Map(RankingsScope value)
+        {
+            switch (value)
+            {
+                case RankingsScope.Performance:
+                    return LayoutStrings.MenuRankingsIndex;
+
+                case RankingsScope.Spotlights:
+                    return LayoutStrings.MenuRankingsCharts;
+
+                case RankingsScope.Score:
+                    return LayoutStrings.MenuRankingsScore;
+
+                case RankingsScope.Country:
+                    return LayoutStrings.MenuRankingsCountry;
+
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(value), value, null);
+            }
+        }
+    }
```
Not sure how valid this is. Existing localised enums have localisations in the same file.

frenzibyte replied:
Probably failed on the wording, but I meant that `RankingsScope` and `RankingsScopeEnumLocalisationMapper` should both move to one `RankingsScope.cs` file.

kj415j45 replied:
This change will go to another PR. won't fix in this one.

frenzibyte replied:
It's a pretty straightforward change so I've done it myself with another fix for the used strings in the mapper (https://github.com/ppy/osu/pull/13914/commits/456f4e6f1fce118405e0947f73705e90d1c30306).

bdach commented:
I'm not sure why the scope of this was expanded to all overlays, going against the advice [given above](https://github.com/ppy/osu/pull/13914#issuecomment-881896001)...

frenzibyte commented:
I failed to notice that comment as the author started to localise everything in this PR, but in the end, I think it's not that complex, it's just a couple of strings in each overlay's header pointing to the right `LocalisableString` in the `Strings` files.

bdach commented:
This needs conflict resolution now.

peppy committed:
Merge branch 'master' into localisation-header

peppy committed:
Remove incorrectly added null coalesce

kj415j45 committed:
Move strings

kj415j45 committed:
Merge https://github.com/kj415j45/osu into localisation-header

kj415j45 committed:
Merge 'master' of ppy/osu into localisation-header

kj415j45 committed:
use same code style

Co-authored-by: frenzibyte <frenzibyte@gmail.com>

kj415j45 committed:
Add localisation for BeatmapSetHeader

kj415j45 committed:
Merge branch 'localisation-header' of https://github.com/kj415j45/osu into localisation-header

frenzibyte committed:
Add `description` to key names of description strings

frenzibyte committed:
Fix incorrect property xmldoc inherited

frenzibyte committed:
Move `RankingsScope` to own file and fix mapper strings

frenzibyte committed:
Fix `WikiHeader` setting `string.Empty` rather than `null` on breadcrumb

That's supposed to be fixed already in the base PR, but somehow it
continued to exist here.

frenzibyte commented:
I've resolved some of the remaining issues I didn't notice on my first review, I think it's in a good state now.

I was hoping the existing title and description strings in `SettingsStrings`/`NotificationsStrings`/etc. to be moved to the new `NamedOverlayComponentStrings`, but very hesitant about doing that now that they're already published to crowdin, so I've left them as they're for now.

This will have to wait for https://github.com/ppy/osu/pull/13918 to conclude, however.

frenzibyte committed:
Fix news/wiki tests checking against incorrect constant string

peppy commented:
This needs conflict resolution too.

frenzibyte committed:
Merge branch 'master' into localisation-header

peppy committed:
Update resources

peppy commented on osu.Game/Localisation/NamedOverlayComponentStrings.cs:
```diff
@@ -0,0 +1,66 @@
+        public static LocalisableString NewsDescription => new TranslatableString(getKey(@"news_description"), @"get up-to-date on community happenings");
+
+        /// <inheritdoc cref="PageTitleStrings.MainWikiControllerDefault"/>
+        public static LocalisableString WikiTitle => PageTitleStrings.MainWikiControllerDefault;
```
Why are these here? Just use the osu-web ones directly, I think?

kj415j45 replied:
`PageTitleStrings` comes from `osu.Game.Resources.Localisation.Web` namespace.

---

and as discussed before, the `*Title` and `*Description` style is for consistency:
```cs
//in components
{
    Title = NamedOverlayComponentStrings./* name here */Title;
    Description = NamedOverlayComponentStrings./* name here */Description;
}
//in osu.Game/Localisation/NamedOverlayComponentStrings.cs
    /// <inheritdoc cref="PageTitleStrings.{string name}"/>
    public static LocalisableString /* name here */Title => PageTitleStrings./* name here */ControllerDefault;

    /// <summary>
    /// "sample string"
    /// </summary>
    public static LocalisableString /* name here */Description => new TranslatableString(getKey(@"_description"), @"sample string");
```

But it's ok using following style :shrug:
```cs
//in components
{
    Title = PageTitleStrings./* name here */ControllerDefault;
    Description = NamedOverlayComponentStrings./* name here */Description;
}
```

peppy replied:
I definitely prefer the direct references. @smoogipoo can you confirm your opinion?

smoogipoo replied:
I also prefer direct references.

kj415j45 replied:
👌🏼 it will be direct references in next commit

peppy replied:
Not sure if it will ever be the case, but I can see these potentially being moved to `osu-web` in the future (if there are more design updates to bring web/client in closer consistency).

frenzibyte replied:
Yeah consistency is really the only reason I've went with that style, but since redirecting in generated strings classes is not a good idea in general, I'm fine with having it use direct references.

Moving the tooltip descriptions web-side would also remove all this hassle and leave things in a consistent state, but that's probably for later when actually needed.

kj415j45 committed:
Use direct reference instead

peppy commented on osu.Game/Overlays/Changelog/ChangelogHeader.cs:
```diff
@@ -21,16 +24,16 @@ public class ChangelogHeader : BreadcrumbControlOverlayHeader
         public ChangelogUpdateStreamControl Streams;
 
-        private const string listing_string = "listing";
+        public static LocalisableString ListingString => LayoutStrings.HeaderChangelogIndex;
```
@smoogipoo thoughts on this one? is used in tests (`public` part of the weirdness) but not sure if we want this, or another case of just inlining the references at each point.

smoogipoo replied:
I think I'm fine with this. I just don't like redirects in the supposed-to-be-autogenerated XStrings classes because the format isn't guaranteed to remain when new strings are added and I can't guarantee it won't break the analyser in the first place.

But this seems okay to me.

peppy commented on osu.Game/Overlays/Changelog/ChangelogHeader.cs:
```diff
@@ -21,16 +24,16 @@ public class ChangelogHeader : BreadcrumbControlOverlayHeader
             Current.ValueChanged += e =>
             {
-                if (e.NewValue == listing_string)
+                if (e.NewValue == ListingString)
```
Genuinely surprised this actually works across language changes..
